clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -163,6 +175,17 @@ public void run() {
      */
     void run(long now) {
         Cluster cluster = metadata.fetch();
+
+        if (transactionState.isIdempotenceEnabled()) {
+            while (!transactionState.pidIsSet() && (time.milliseconds() - now) < maxBlockTimeMs) {
```

Comment:
This makes sense. To capture our offline discussion: these synchronous background RPC's will try indefinitely. These include the initPidRequest, and all the transactional RPCs. However, the async produce requests will queue behind these synchronous RPC's, and eventually the send will timeout and return an error to the user.